### ISSUE TITLE
Unable to communicate with Edge modules (in EFLOW) from the host OS

### ISSUE THREAD

**Issue by bhjertaas** [Azure/iotedge#7220]
## Expected Behavior
We should be able to communicate with Edge modules from the host OS 

## Current Behavior
Sometimes, after reboot of the PC, communication from host OS to Edge modules does not work. We can `Connect-EflowVm` and `Get-EflowVmAddr` and things, but making HTTP calls to the Edge modules within Eflow doesn't work. The requests simply time out. We don't have much exceptions and logs to provide unfortunately, but here is the output from a Powershell on Windows

```
PS C:\Windows\System32> get-EflowVmAddr 
[01/17/2024 11:07:07] Querying IP and MAC addresses from virtual machine (DESKTOP-E4VKT5P-EFLOW)
  - Virtual machine MAC: 00:15:5d:43:7d:b6
 - Virtual machine IP : 172.18.92.180 retrieved directly from virtual machine
 00:15:5d:43:7d:b6
172.18.92.180 

PS C:\Windows\System32> curl http://172.18.92.180:9602/metrics curl: (28) Failed to connect to 172.18.92.180 port 9602 after 21008 ms: Couldn't connect to server 
PS C:\Windows\System32> curl http://172.18.92.180:1337/metadata curl: (28) Failed to connect to 172.18.92.180 port 1337 after 21031 ms: Couldn't connect to server 
```

The following image shows the problem over time. If the PC is unable to make simple HTTP requests the graph drops to zero.  The test-PCs are setup to restart every 3 hours.
![image](https://github.com/Azure/iotedge/assets/22844712/f75d876b-5e67-4d14-945f-791bab36f1a3)

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. When this happens, one can test it by running `Get-EflowVmAddr` and copy the IP address that comes out
2. Try to make a request to a module like this `curl http://172.18.92.180:9602/metrics`
which in our case gets metric data from edgeHub because we have mapped the internal 9600 port to 9602. But it doesn't matter what you request, the request won't work.

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```

Configuration checks (aziot-identity-service)
---------------------------------------------
√ keyd configuration is well-formed - OK
√ certd configuration is well-formed - OK
√ tpmd configuration is well-formed - OK
√ identityd configuration is well-formed - OK
√ daemon configurations up-to-date with config.toml - OK
√ identityd config toml file specifies a valid hostname - OK

‼ aziot-identity-service package is up-to-date - Warning
    Installed aziot-identity-service package has version 1.4.6 but 1.4.7 is the latest stable version available.
    Please see https://aka.ms/aziot-update-runtime for update instructions.
√ host time is close to reference time - OK
√ preloaded certificates are valid - OK
√ keyd is running - OK
√ certd is running - OK
√ identityd is running - OK
√ read all preloaded certificates from the Certificates Service - OK
√ read all preloaded key pairs from the Keys Service - OK
√ check all EST server URLs utilize HTTPS - OK
√ ensure all preloaded certificates match preloaded private keys with the same ID - OK

Connectivity checks (aziot-identity-service)
--------------------------------------------
‼ host can connect to and perform TLS handshake with iothub AMQP port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub HTTPS / WebSockets port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub MQTT port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
√ host can connect to and perform TLS handshake with DPS endpoint - OK

Configuration checks
--------------------
√ aziot-edged configuration is well-formed - OK
√ configuration up-to-date with config.toml - OK
√ container engine is installed and functional - OK
√ configuration has correct URIs for daemon mgmt endpoint - OK
‼ aziot-edge package is up-to-date - Warning
    Installed IoT Edge daemon has version 1.4.20 but 1.4.27 is the latest stable version available.
    Please see https://aka.ms/iotedge-update-runtime for update instructions.
√ container time is close to host time - OK
√ DNS server - OK
√ production readiness: logs policy - OK
√ production readiness: Edge Agent's storage directory is persisted on the host filesystem - OK
√ production readiness: Edge Hub's storage directory is persisted on the host filesystem - OK
√ proxy settings are consistent in aziot-edged, aziot-identityd, moby daemon and config.toml - OK

Connectivity checks
-------------------
26 check(s) succeeded.
5 check(s) raised warnings. Re-run with --verbose for more details.
7 check(s) were skipped due to errors from other checks. Re-run with --verbose for more details.
iotedge-user@DESKTOP-5SO2V9I-EFLOW [ ~ ]$ sudo iotedge check

Configuration checks (aziot-identity-service)
---------------------------------------------
√ keyd configuration is well-formed - OK
√ certd configuration is well-formed - OK
√ tpmd configuration is well-formed - OK
√ identityd configuration is well-formed - OK
√ daemon configurations up-to-date with config.toml - OK
√ identityd config toml file specifies a valid hostname - OK
‼ aziot-identity-service package is up-to-date - Warning
    Installed aziot-identity-service package has version 1.4.6 but 1.4.7 is the latest stable version available.
    Please see https://aka.ms/aziot-update-runtime for update instructions.
√ host time is close to reference time - OK
√ preloaded certificates are valid - OK
√ keyd is running - OK
√ certd is running - OK
√ identityd is running - OK
√ read all preloaded certificates from the Certificates Service - OK
√ read all preloaded key pairs from the Keys Service - OK
√ check all EST server URLs utilize HTTPS - OK
√ ensure all preloaded certificates match preloaded private keys with the same ID - OK

Connectivity checks (aziot-identity-service)
--------------------------------------------
‼ host can connect to and perform TLS handshake with iothub AMQP port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub HTTPS / WebSockets port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub MQTT port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
√ host can connect to and perform TLS handshake with DPS endpoint - OK

Configuration checks
--------------------
√ aziot-edged configuration is well-formed - OK
√ configuration up-to-date with config.toml - OK
√ container engine is installed and functional - OK
√ configuration has correct URIs for daemon mgmt endpoint - OK
‼ aziot-edge package is up-to-date - Warning
    Installed IoT Edge daemon has version 1.4.20 but 1.4.27 is the latest stable version available.
    Please see https://aka.ms/iotedge-update-runtime for update instructions.
√ container time is close to host time - OK
√ DNS server - OK
√ production readiness: logs policy - OK
√ production readiness: Edge Agent's storage directory is persisted on the host filesystem - OK
√ production readiness: Edge Hub's storage directory is persisted on the host filesystem - OK
√ proxy settings are consistent in aziot-edged, aziot-identityd, moby daemon and config.toml - OK

Connectivity checks
-------------------
26 check(s) succeeded.
5 check(s) raised warnings. Re-run with --verbose for more details.
7 check(s) were skipped due to errors from other checks. Re-run with --verbose for more details.

```
</details>

### Device Information
* Host OS [e.g. Ubuntu 22.04, Windows Server IoT 2019]: 1.4.10.25103 on Windows 11
* Architecture [e.g. amd64, arm32, arm64]:  amd64
* Container OS [e.g. Linux containers, Windows containers]: Mariner Linux (in Eflow)

### Runtime Versions
* aziot-edged [run `iotedge version`]: 1.4.20
* Edge Agent [image tag (e.g. 1.0.0)]: mcr.microsoft.com/azureiotedge-agent:1.4
* Edge Hub [image tag (e.g. 1.0.0)]: mcr.microsoft.com/azureiotedge-hub:1.4
* Docker/Moby [run `docker version`]: 20.10.27 (iotedge_moby_engine version -> 24.0.6)

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Additional Information

When we do encounter this issue, it remains a problem until the PC is restarted.

- When we go into the VM with connect-eflowVm and run `sudo iptables -P input accept` we're able to ping the EFLOW VM from the host OS. 
- Running hcsdiag list to get the GUID, and then `hcsdiag console <guid>` we can confirm that from that shell, we can both ping the IP of the VM and do curl requests to edge module endpoints that did not work from the host OS

We have been in contact with the Iotedge-eflow team, and one member of that team have had a session with us on one of the test PCs that did not work. He believes the problem has to do with the IoT-Edge part, which is why I post this issue here.

**Comment by nyanzebra** [comment]
@bhjertaas would you mind providing a support bundle? The command is `iotedge support-bundle`. 

From what I understand, there is no issue retrieving the metrics via curl from EFLOW vm side, but when trying from host machine side it does not work. Is this correct?

Additionally, is this a recent issue? Or something that has been happening?

**Comment by nyanzebra** [comment]
@jagadishmurugan while edge team is investigating the support bundle is there anything to recommend to @bhjertaas to validate EFLOW to host networking. Just want to make sure that all looks correct.

**Comment by konichi3** [comment]
@jagadishmurugan @nyanzebra Can you follow up and share your finding?

**Comment by konichi3** [comment]
@jagadishmurugan @nyanzebra Any updates on this?

**Comment by nyanzebra** [comment]
@bhjertaas sorry for delay, this got lost amongst other issues I have been looking at. The support bundle provided seems to have failed to get logs `Error grabbing logs: log message is too large (973735539 > 1000000)`. Would it be possible to run a smaller window of time support bundle? For example `support-bundle --since 6h`?

**Comment by bhjertaas** [comment]
I have tried to make a support-bundle with smaller time frame, but the "log message is too large" problem remains. 

Until the underlying [Moby bug](https://github.com/moby/moby/issues/46699) is fixed, the recommended _Local_ logging driver is not a viable option. Which [log driver](https://docs.docker.com/config/containers/logging/configure/) do you recommend instead on Edge devices? I'm just thinking that perhaps we will have to sort that out first, before we can get clarity on this issue.

**Comment by nyanzebra** [comment]
@bhjertaas sorry for the delay here, from the linked Moby bug it seems like there is no known solution for the logger at this time. By default I believe the local log driver is used: https://learn.microsoft.com/en-us/azure/iot-edge/production-checklist?view=iotedge-1.4#set-up-default-logging-driver

Beyond that @veyalla, do we have any other official stance on log drivers? I believe we leave it open to customer needs as in above doc and that is it. Just want to double confirm.

**Comment by veyalla** [comment]
If there is no line of sight from Moby side to the fix the `local` log driver. I'd recommend we change our docs to suggest `json` w/ size limits as the default log driver. Adding @bishal41 @PatAltimore for following up on docs change.

**Comment by bishal41** [comment]
@veyalla - I was chatting with @cpuguy83 and changing our docs to use json-file logging driver as a default won't fix the issue as it's the same problem with both drivers. @bhjertaas would you be able to try the suggestions provided in this issue and see if that helps? https://github.com/Azure/iotedge/issues/7074#issuecomment-2062913691

**Comment by mmuskala-c4workx** [comment]
Hi @bhjertaas did you ever get to fix the original issue? We're facing similar problem.

**Comment by benny-cowboy** [comment]
I am no longer working on that project, but from memory it kept being an issue. I think converting to Linux as the host OS is the only feasible solution.

**Comment by mmuskala-c4workx** [comment]
> I am no longer working on that project, but from memory it kept being an issue. I think converting to Linux as the host OS is the only feasible solution.

Ok, thanks for the reply!
